server/src/internalClusterTest/java/org/elasticsearch/snapshots/CorruptedBlobStoreRepositoryIT.java

```diff
@@ -760,28 +763,81 @@ public void testSnapshotWithMissingShardLevelIndexFile() throws Exception {
             Files.delete(shardGen);
         }
 
-        logger.info("--> creating another snapshot");
+        if (randomBoolean()) {
+            logger.info(
+                "--> restoring the snapshot, the repository should not have lost any shard data despite deleting index-N, "
+                    + "because it uses snap-*.data files and not the index-N to determine what files to restore"
+            );
+            indicesAdmin().prepareDelete("test-idx-1", "test-idx-2").get();
+            RestoreSnapshotResponse restoreSnapshotResponse = clusterAdmin().prepareRestoreSnapshot(
+                TEST_REQUEST_TIMEOUT,
+                "test-repo",
+                "test-snap-1"
+            ).setWaitForCompletion(true).get();
+            assertEquals(0, restoreSnapshotResponse.getRestoreInfo().failedShards());
+            ensureGreen("test-idx-1", "test-idx-2");
+        }
+
+        logger.info("--> creating another snapshot, which should re-create the missing file");
+        try (
+            var ignored = new BlobStoreIndexShardSnapshotsIntegritySuppressor();
+            var mockLog = MockLog.capture(BlobStoreRepository.class)
+        ) {
+            mockLog.addExpectation(
+                new MockLog.SeenEventExpectation(
+                    "fallback message",
+                    "org.elasticsearch.repositories.blobstore.BlobStoreRepository",
+                    Level.ERROR,
+                    "shard generation [*] in [test-repo][*] not found - falling back to reading all shard snapshots"
+                )
+            );
+            mockLog.addExpectation(
+                new MockLog.SeenEventExpectation(
+                    "shard blobs list",
+                    "org.elasticsearch.repositories.blobstore.BlobStoreRepository",
+                    Level.ERROR,
+                    "read shard snapshots [*] due to missing shard generation [*] in [test-repo][*]"
+                )
+            );
+            CreateSnapshotResponse createSnapshotResponse = clusterAdmin().prepareCreateSnapshot(
+                TEST_REQUEST_TIMEOUT,
+                "test-repo",
+                "test-snap-2"
+            ).setWaitForCompletion(true).setIndices("test-idx-1").get();
+            assertEquals(
+                createSnapshotResponse.getSnapshotInfo().totalShards(),
+                createSnapshotResponse.getSnapshotInfo().successfulShards()
+            );
+            mockLog.assertAllExpectationsMatched();
```

Comment:
Should we also check that a new shard generation file is written?

Reply:
++ b6434df7c220099f4e383ddea1e5b7ff850544db